fuzz/fuzz_targets/native_c.rs

```diff
@@ -0,0 +1,140 @@
+#![no_main]
+
+use std::ffi::{CStr, CString};
+use std::fmt::Write;
+
+use libfuzzer_sys::fuzz_target;
+
+#[derive(Debug)]
+enum State<'a> {
+    Ok(&'a str),
+    Overflow,
+}
+
+pub struct FormatBuf<'a> {
+    buf: &'a mut String,
+    max_len: usize,
+}
+
+impl<'a> Write for FormatBuf<'a> {
+    fn write_str(&mut self, s: &str) -> Result<(), std::fmt::Error> {
+        if self.buf.len() + s.len() > self.max_len {
+            return Err(std::fmt::Error);
+        }
+        self.buf.push_str(s);
+        Ok(())
+    }
+}
+
+fn asciify(x: &str) -> String {
+    let mut result = String::with_capacity(x.len() * 4);
+    for ch in x.chars() {
+        if ch.is_ascii() {
+            result.push(ch);
+        } else {
+            write!(&mut result, "\\u{{{:x}}}", ch as u32).ok();
+        }
+    }
+    result
+}
+
+fuzz_target!(|data: &[u8]| {
+    if data.len() == 0 {
```

Comment:
Can we add a test to make sure an empty input is OK too?

Reply:
will add